### PR TITLE
[otbn] RTL implementation of the KMAC interface

### PR DESCRIPTION
**Pull request by etterli**
The last commit of this PR implements the OTBN side of the KMAC interface. The interface is specified in [#30278](https://github.com/lowRISC/opentitan/pull/30278).

All other commits are part of one of these PRs:
- https://github.com/lowRISC/opentitan/pull/30141: RTL on KMAC side
- https://github.com/lowRISC/opentitan/pull/30143: DV for KMAC side
- https://github.com/lowRISC/opentitan/pull/30278: Specification of this implementation.

### CODE REVIEW

**Comment by etterli** [comment]
The interface no longer automatically sends a process command when a partial message is sent. It just detects it and waits for SW to issue a process command.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv [hunk 1]
```diff
@@ -0,0 +1,1035 @@
...
+    if (!rst_ni) begin
+      sec_wipe_detected_q <= '0;
+    end else begin
+      sec_wipe_detected_q <= sec_wipe_detected_d;
+    end
+  end
+
+  ////////////////////////////////////
+  // KMAC_DATA_Sx secure wipe delay //
+  ////////////////////////////////////
```
Comment (etterli) [inline hw/ip/otbn/rtl/otbn_kmac_if.sv]: This whole concept of delaying the secure wipe for the relevant 64 bits (or rather 128 for both shares) of the message is quite ugly. But I haven't found a better way how to ensure that the data does not change. And there is an assertion in the msg FIFO inside the KMAC (prim packer and also downstream) that it must remain stable.

Any idea is welcome.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 2]
```diff
@@ -0,0 +1,1035 @@
...
+  assign ispr_kmac_data_s1_wdata = ispr_kmac_data_s1_wdata_i;
+
+  // MUX the KMAC response with a write instruction. The response has priority over the WSR write.
+  // Any secure wipe has the highest priority. The wipe must be handled on word level due to the
+  // pending wipe mechanism.
+  for (genvar word = 0; word < BaseWordsPerWLEN; word++) begin : g_wsr
+    // Detect whether the wipe for this word was delayed. The message beat index is frozen during a
+    // secure wipe. Words 0,1 relate to beat 0, words 2,3 to beat 1, etc.
+    logic is_prevented_word;
+    assign is_prevented_word = MsgPartsW'(word / 2) == msg_beat_idx;
```
Comment (etterli) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: This currently fails the linting because we compare signed with unsigned. Currently looking for a better solution.

Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Is now fixed but still, the whole concept is not super clean.. Let me know if there is a better solution.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv [hunk 3]
```diff
@@ -0,0 +1,1040 @@
...
+  assign ispr_kmac_ctrl_r = '{
+    rsvd2: '0,
+    cfg:   kmac_cfg_q,
+    rsvd1: '0,
+    cmd:   '0
+  };
+  assign ispr_kmac_ctrl_rdata_o = ispr_kmac_ctrl_r;
+
+  // Extract command from write.
+  assign current_cmd = ispr_kmac_ctrl_w.cmd & {$bits(kmac_cmd_t){ispr_kmac_ctrl_wr_i}};
```
Comment (etterli) [inline hw/ip/otbn/rtl/otbn_kmac_if.sv:700]: This extracts the cmd from the current instruction and thus makes `current_cmd` combinatorial. I designed the FSM in such a way that there is no timing path directly to the interface. But maybe it would make more sense if the write goes to an actual register and the FSM decides on a flopped value?

File: hw/ip/otbn/rtl/otbn_kmac_if.sv [hunk 4]
```diff
@@ -0,0 +1,1040 @@
+// Copyright lowRISC contributors (OpenTitan project).
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac_if.sv:1]: For consistency, this file and module should be named `otbn_kmac_if` (see other blocks with KMAC IF such as keymgr, lc_ctrl, ...).

And the file should be added to the .core file.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 5]
```diff
@@ -0,0 +1,1040 @@
...
+
+  typedef enum logic [1:0] {
+    ReqDataCfg = 2'b00,
+    ReqDataMsg = 2'b01,
+    ReqDataOff = 2'b10
+  } otbn_kmac_req_data_sel_e;
+
+  typedef struct packed {
+    logic done;
+    logic prc; // process - is a reserved keyword
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: proc?

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 6]
```diff
@@ -0,0 +1,1040 @@
...
+    logic rsp_error;
+    logic rsp_valid;
+    logic msg_write_error;
+    logic ctrl_error;
+    logic ready;
+  } kmac_status_t;
+
+  // TODO: otbn_base_intg_word_t and kmac_ispr_urnd_t are the same as in MAI.
+  // How can we combine these? Should we move this to the otbn_pkg? Could also define it properly
+  // in otbn_pkg and use it everywhere in OTBN.
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: Yes, those should be moved to otbn_pkg.sv then.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv [hunk 7]
```diff
@@ -0,0 +1,1040 @@
...
+          // happened. We must send the start request if it was already pending, otherwise a valid
+          // would be dropped. We do not track if a request was pending, so always send it.
+          state_d = sec_wipe_detected ? OtbnKmacProcessing : OtbnKmacWaitForMsg;
+        end
+      end
+      OtbnKmacWaitForMsg: begin
+        if (sec_wipe_detected) begin
+          // Immediately end the message phase.
+          state_d = OtbnKmacProcessing;
+        end else if (current_cmd.send && !no_more_msg_allowed_q) begin
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac_if.sv:278]: Please add some comments here.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 8]
```diff
@@ -0,0 +1,1040 @@
...
+  logic msg_beat_is_partial;
+  logic msg_beat_is_empty;
+  logic is_last_msg_beat;
+  logic msg_beat_sent;
+  logic end_msg_phase;
+
+  // Request data and strobe selection
+  always_comb begin
+    req_data_s0 = '0;
+    req_data_s1 = '0;
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: For area and security reasons (FI) I would reduce the muxing here. Please forward always req_msg_s0/s1 unless when sending the command/config. Then, we should probably mux together the relvant command/config bits + the bits of req_msg_s0/s1 for the bits which are not used for the command. This way it's harder to inject deterministic data. At most an adversary could then glitch the command bits to something deterministic, but not the entire data beat.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 9]
```diff
@@ -0,0 +1,1040 @@
...
+      default: begin
+        // Same as off
+        req_data_s0 = '0;
+        req_data_s1 = '0;
+        req_strb    = '0;
+      end
+    endcase
+  end
+
+  // A counter to step through the message words making up the complete message in the WSRs.
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: message words making up the complete message in the WSRs -> message beats of the current message part in the WSRs

**Review by vogelpi — COMMENTED** [review]
Thanks Pascal, this looks pretty nice so far. I've reviewed the first half and will continue later.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 4 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:1]: Ok, will rename it.

There will be a 2nd PR for integrating the kmac interface into OTBN. Then I will add it to the .core file.

If you want, I can already open a PR for this? (If it helps to review)

File: hw/ip/otbn/rtl/otbn_kmac_if.sv [hunk 10]
```diff
@@ -0,0 +1,1040 @@
...
+  assign ispr_kmac_data_s1_wdata = ispr_kmac_data_s1_wdata_i;
+
+  // MUX the KMAC response with a write instruction. The response has priority over the WSR write.
+  // Any secure wipe has the highest priority. The wipe must be handled on word level due to the
+  // pending wipe mechanism.
+  for (genvar word = 0; word < BaseWordsPerWLEN; word++) begin : g_wsr
+    // Detect whether the wipe for this word was delayed. The message beat index is frozen during a
+    // secure wipe. Words 0,1 relate to beat 0, words 2,3 to beat 1, etc.
+    logic is_prevented_word;
+    assign is_prevented_word = MsgPartsW'(unsigned'(word) / 2) == msg_beat_idx;
```
Comment (etterli) [inline hw/ip/otbn/rtl/otbn_kmac_if.sv]: The whole concept of cleaning just a word is not super clean.. Let me know if there is a better solution.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 5 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: renamed it

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 6 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Created https://github.com/lowRISC/opentitan/pull/30346 to do this. The relevant commit is now also in this PR until it is merged.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 7 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:278]: Explained it.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 4 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:1]: Ah no, that's okay. Don't worry then.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 8 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Good idea. I reworked it accordingly. Used your nice explanation as comment :)

File: hw/ip/otbn/rtl/otbn_kmac_if.sv [hunk 11]
```diff
@@ -0,0 +1,1040 @@
...
+
+  // Start command is sent once handshake occurred.
+  assign start_req_sent = send_start_req & app_rsp_i.req_ready;
+
+  // Sending a message consists of sending 64-bit message parts in up to 4 message beats.
+  // The last message beat can be partial. OTBN does not set the last flag in this case. OTBN SW
+  // must always end the message phase by issuing a PROCESS command which will send an empty
+  // message with the last flag asserted.
+  // If the full 256-bit message is partial but a multiple of 64 bits, then some beats are empty
+  // messages. These will not be sent. Instead of looking ahead, we just detect whether the current
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac_if.sv]: You already wrote above that there can be up to 4 message beats. So this can be shorter here, e.g.: "If the current message beat is empty, we do not place a message request and end the sending."

The last sentence can also be simpler: "If a secure wipe happens, we only send the current message beat and then end the sending."

**Review by etterli — COMMENTED** [review]
Thanks @vogelpi for the first review. Here is the updated version. The KMAC_CTRL is now split into KMAC_CTRL and KMAC_CFG.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 12]
```diff
@@ -0,0 +1,1040 @@
...
+  logic discarded_rsp_hs;
+
+  // Accept the next response when FSM tells so and if the current response has been consumed.
+  // This is based upon the registered valid flag to avoid that the current instruction drives the
+  // ready which would lead to a timing path from insn decoding to the KMAC interface. It however
+  // means that the next digest data is available two cycles after the data is read. The sequence
+  // thus is (-> means next cycle):
+  // Insn reads data, valid is reset -> interface accepts next digest -> New digest available
+  // TODO: Is this performance impact ok? Do we have to cut the interface so we can immediately
+  //       accept the next data based upon data_consumed?
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: Yes, we should cut this. My understanding is that receiving the digest is far more critical then sending the message (see previous messages). I think it should be possible to break this timing path by using the predecoder, no?

Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Basically, we want:
-> register digest_s0/s1
-> ispr_kmac_data_s0_rd_i -> ispr_kmac_data_s0_rd_i + register digest_s0/s1
-> ispr_kmac_data_s0_rd_i -> ispr_kmac_data_s0_rd_i + register digest_s0/s1

So on average, we should be able to read the two ISPRs every two clock cycles.

For this, the data_rsp_ready needs to depend on ispr_kmac_data_s0_rd_i || ispr_kmac_data_s1_rd_i . Does this make sense?

Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Maybe it helps to look at this from the KMAC side: KMAC wants to push with the maximum bandwidth. On the OTBN side, we know what this bandwidth is and once we've seen the first digest valid bit, we want to just read everything and we can do that blindly. We don't care anymore about polling the valid bit and then reacting to that.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 13]
```diff
@@ -0,0 +1,1040 @@
...
+  // Extract command from write.
+  assign current_cmd = ispr_kmac_ctrl_w.cmd & {$bits(kmac_cmd_t){ispr_kmac_ctrl_wr_i}};
+
+  // Capture writes to configuration bits.
+  assign kmac_cfg_d = wipe_configuration  ? '0                   :
+                      ispr_kmac_ctrl_wr_i ? ispr_kmac_ctrl_w.cfg :
+                                            kmac_cfg_q;
+
+  // TODO: Do we need integrity or any other duplication? We could duplicate the config and then
+  // send the config on both data share signals.
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: Yes, I would just do that. And I would duplicate these bits also in the CSR. So software would have to provide the config for both shares. These couple of flops are not expensive but it's a lot simpler from a hardening perspective than reasoning about where what could go wrong.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 14]
```diff
@@ -0,0 +1,1040 @@
...
+  assign app_cfg = '{
+    en_xof:      kmac_cfg_q.en_xof,
+    kstrength:   kmac_cfg_q.kstrength,
+    mode:        kmac_cfg_q.mode,
+    prefix_mode: 1'b0 // not relevant
+  };
+
+  // KMAC_STRB
+  // TODO: Do we need integrity?
+  assign ispr_kmac_strb_rdata_o = kmac_strb_q;
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: I think it should be sufficient to check on the receiving side, i.e., inside KMAC that there are no empty message beats (strobe == 0) in between the last beat and the process command. I am not worried about dropping or shortening a single beat, but more about sending all-zero messages.

File: hw/ip/otbn/rtl/otbn_kmac.sv [hunk 15]
```diff
@@ -0,0 +1,1040 @@
...
+  // TODO: This can probably also be done with regular FFs. Would this be better?
+  always_ff @(posedge clk_i) begin
+    pending_data_s0_urnd_q <= pending_data_s0_urnd_d;
+    pending_data_s1_urnd_q <= pending_data_s1_urnd_d;
+  end
+
+  ////////////////
+  // WSR access //
+  ////////////////
+  // The upper 6 words always read as zero.
```
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_kmac.sv]: This effectively prevents reading back the data registers after having written them (to pass data to KMAC). This may hurt us in the future in case this would be required for hardening.

Can't we just always read all data back but the digest response just overwrites the lowest 64 bits?

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 10 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv]: My gut feeling is that this is an optimization that we don't really need. It's really an extreme corner case and corner cases can be slow.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 1 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv]: In my view we don't need to buffer / delay the secure wipe word here for the following reasons:
- The secure wipe is fed from URND and we do two wipes with a URND reseed in between.
- We do two wipes with a reseed in between to not allow recovering the URND output of the first wipe as this could potentially allow for a SCA attack on the secret using the leakage between the secret and the URND output.
- The secret here is **always** shared because KMAC consumes and provides shared data. So there would be at most second order leakage. This is different from regular WDRs where we can have shared and unshared data.

So instead of buffering / delaying the wipe I would do the following:
- If there is not outstanding request, wipe based on the wipe input signal and using URND.
- If there is an outstanding request
  - Track that we have to do the wipe at a later point.
  - Once that point comes, generate a URND update/advance signal (no reseed). At that point, the main wipe may have finished and OTBN may be idle (URND at rest) otherwise.
- Wipe the entire WSRs instead of individual 64-bit beats to reduce complexity.

Would that work?

Alternatively, I would also be fine to not adhere to the valid-locked-in principle here. This is really a corner case and I am not sure we will be able to trigger this at the top level. We may be able to trigger it with a FI simulation test but then can also deactivate the relevant SVAs in KMAC and reason why this is necessary for that specfiic test.

**Review by vogelpi — COMMENTED** [review]
I've now also reviewed the rest. I would try to simplify the secure wipe (it's a corner case) and instead optimize the digest reading (the fast case we care about).

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 1 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv]: Thank you for the proposal. This sounds good. Not guaranteeing two wipes for these WSRs just means that these should not be used as extra WDRs when the KMAC IF is not used.

I also checked the consequences of not adhering to the valid locked-in principle again in more detail. If it is violated, then the only consequence is that the digest value computed by the KMAC is undefined. But it does not lock up the KMAC IP. The  recovery mechanism will still bring the KMAC IP back to the Idle state. And when a secure wipe happens, we anyway ignore any digest data. The strobe however should be stable.

Therefore, I think the best is if we just secure wipe these registers in the normal way and accept the corner case. The configuration and strobe registers can simply be reset to the default values once the recovery ends.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 10 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv]: This is now obsolete with the secure wipe rework.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 13 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Ok if we postpone this because it also requires changes on the KMAC side? I created an issue for this. Rephrased the TODO so it's no longer a question.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 14 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Manipulating the strobe can result in undefined behaviour / digest values or a message is interpreted as empty message. Both cases must be detected somehow. I agree that this should be handled on the KMAC side because otherwise we would also have to harden the data signals.

Created an issue for this.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 15 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Sure we can, that makes sense. And if SW requires to clear the WSRs before (or between) receiving digest data it can always write to it first.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 12 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Ah thanks for reminding me that the relevant read control signals for the KMAC_DATA_Sx WSRs are predecoded. This resolves any concerns regarding the timing.

I have now redesigned it such that we can accept the next response in the same cycle when SW reads the previous digest.

**Comment by etterli** [comment]
The latest version features:
- Rebased on the newest version of https://github.com/lowRISC/opentitan/pull/30143 (minimized DV changes and squashed everything).
- The implementation of the new OTBN SW command `CLOSE`
- Simplified the secure wipe behaviour for `KMAC_DATA_S0/1`.
  - These are now wipe like any other WSR.
  - We accept the corner case that we violate the valid locked-in principle for the request channel when a wipe happens whilst sending a message.
- The interface is now capable of accepting the next response in the same cycle the previous digest is consumed/read.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 1 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv]: Fully agree, thanks @etterli !

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 7 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:278]: Thanks a lot, this is great!

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 8 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Thanks that's wonderful. But maybe it's easier to write this as a always_comb statement:
```sv
always_comb begin
  req_data_s0 = req_msg_s0;
  req_data_s1 = req_msg_s1;

  if (ReqDataCfg) begin
   req_data_s0[$bits(kmac_pkg::app_ses_config_t) - 1:0] = app_cfg;
   req_data_s1[$bits(kmac_pkg::app_ses_config_t) - 1:0] = app_cfg;
  end
end
```

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 12 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Fantastic!

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 13 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Actually, it doesn't require KMAC side changes. KMAC should ignore share1 for config commands, no?

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 15 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Thank you.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 3 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:700]: Hm, good question. If we used a CSRWI instruction, the command would come out of an immediate but also other paths would be possible. Maybe we need a 1 cycle delay then for the START command. This shouldn't be critical for performance as it's done just once per session.

**Review by vogelpi — APPROVED** [review]
Very nice, thanks for the various improvements @etterli !

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 3 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:700]: We now have already a one cycle delay because the FSM will issue the request in another state than when it detects the command (`OtbnKmacIdle` vs `OtbnKmacStarting`, and similar for all other commands). In `OtbnKmacIdle` the state transition depends on the command. And in `OtbnKmacStarting` the FSM sets the request valid signal but the command signal is not evaluated anymore. 

The question now is whether having 2 states actually serves as a cut between the command signal and the request signal generation. This comes down how the FSM is implemented/optimized when synthesizing the design. But I suggest let's do a synthesis and see the effect. This would be easy to refactor.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 13 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: Ah yes true. It is now implemented on the OTBN side.

File: hw/ip/otbn/rtl/otbn_kmac.sv (hunk 8 above)
Comment (etterli) [reply hw/ip/otbn/rtl/otbn_kmac.sv]: This is indeed cleaner :D Changed it.

**Review by nasahlpa — APPROVED** [review]
Thanks Pascal - I have checked whether the RTL matches the spec, which is the case.

**Review by etterli — COMMENTED** [review]
I have now addressed all the feedback and implemented the duplicated `KMAC_CFG` register.

Also see https://github.com/lowRISC/opentitan/pull/30384 for how the interface is integrated into OTBN.

File: hw/ip/otbn/rtl/otbn_kmac_if.sv (hunk 3 above)
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_kmac_if.sv:700]: Yes, I fully agree with you here.

**Review by vogelpi — APPROVED** [review]
Thanks, this looks really great @etterli !

**Comment by vogelpi** [comment]
CHANGE AUTHORIZED: hw/ip/otbn/rtl/otbn_kmac_if.sv

**Comment by nasahlpa** [comment]
CHANGE AUTHORIZED: hw/ip/otbn/rtl/otbn_kmac_if.sv

**Comment by nasahlpa** [comment]
merging as the Private CI run already passed but it is not correctly visualized here and the time out in the FPGA test run is fine.
